pkg/state/cluster/consul/consul.go

```diff
@@ -0,0 +1,187 @@
+package consul
+
+import (
+	"encoding/json"
+	"strings"
+	"time"
+
+	"github.com/gofrs/uuid"
+	"github.com/hashicorp/consul/api"
+	"github.com/jrasell/sherpa/pkg/client"
+	"github.com/jrasell/sherpa/pkg/state"
+	"github.com/jrasell/sherpa/pkg/state/cluster"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+)
+
+const (
+	sessionLockName   = "Sherpa Lock"
+	clusterInfoPath   = "cluster/info"
+	clusterLockPath   = "cluster/lock"
+	clusterLeaderPath = "cluster/leader/"
+)
+
+type ClusterBackend struct {
+	client *api.Client
+	kv     *api.KV
+	logger zerolog.Logger
+
+	clusterInfoPath   string
+	clusterLockPath   string
+	clusterLeaderPath string
+
+	sessionTTL   string
+	lockWaitTime time.Duration
+}
+
+type ClusterLock struct {
+	client *api.Client
+	key    string
+	lock   *api.Lock
+}
+
+func NewStateBackend(log zerolog.Logger, path string) cluster.Backend {
+	consul, _ := client.NewConsulClient()
+
+	return &ClusterBackend{
+		client:            consul,
+		kv:                consul.KV(),
+		clusterInfoPath:   path + clusterInfoPath,
+		clusterLockPath:   path + clusterLockPath,
+		clusterLeaderPath: path + clusterLeaderPath,
+		logger:            log,
+		sessionTTL:        api.DefaultLockSessionTTL,
+		lockWaitTime:      api.DefaultLockWaitTime,
+	}
+}
+
+func (c ClusterBackend) DeleteLeaderEntries(uuid uuid.UUID) {
+	keys, _, err := c.kv.Keys(c.clusterLeaderPath, "/", nil)
+	if err != nil {
+		c.logger.Error().Err(err).Msg("failed to list leader entries")
+		return
+	}
+
+	for _, val := range keys {
+		id := strings.TrimPrefix(val, c.clusterLeaderPath)
+		if id != uuid.String() {
+			_, err := c.kv.Delete(val, nil)
+			if err != nil {
+				c.logger.Error().Err(err).Msg("failed to delete leadership entry")
+			}
+		}
+	}
+}
+
+func (c ClusterBackend) DeleteLeaderEntry(uuid uuid.UUID) error {
+	key := c.clusterLeaderPath + uuid.String()
+	_, err := c.kv.Delete(key, nil)
+	return err
+}
+
+func (c ClusterBackend) GetClusterInfo() (*state.ClusterInfo, error) {
+	kv, _, err := c.kv.Get(c.clusterInfoPath, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if kv == nil {
+		return nil, nil
+	}
+
+	info := &state.ClusterInfo{}
+	if err := json.Unmarshal(kv.Value, info); err != nil {
+		return nil, err
+	}
+	return info, nil
+}
+
+func (c ClusterBackend) PutClusterInfo(info *state.ClusterInfo) error {
+	bytes, err := json.Marshal(info)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.kv.Put(&api.KVPair{Key: c.clusterInfoPath, Value: bytes}, nil)
+	return err
+}
+
+func (c ClusterBackend) GetClusterLeader(id string) (*state.ClusterMember, error) {
+	kv, _, err := c.kv.Get(c.clusterLeaderPath+id, nil)
```

Comment:
Should there ba a `/` between these path components?

Reply:
the const is defined as should I think we are good here: `clusterLeaderPath = "cluster/leader/"`

Reply:
My bad missed the trailing slash.